### PR TITLE
various typing fixes

### PR DESCRIPTION
**Pull request by arvidn**
This PR adds type stubs for `blspy` and addresses mypy errors that resulted from it.

There are 3 lines missing coverage:

```
chia/full_node/full_node.py (0.0%): Missing lines 1297
```
This is a trivially correct change (I think). It's encoding an invariant we have in FullBlock as an `assert`.
This function seems simple to break out and unit test, which we should do in the future.

```
chia/rpc/wallet_rpc_api.py (0.0%): Missing lines 2257
```
This is a typo in the DID recovery code. It should have a test. Matt reviewed the change and said it looks good. I think it's a fairly trivially correct change.

```
tests/util/benchmark_cost.py (0.0%): Missing lines 116
```
We don't instrument benchmarks for coverage. And we probably don't even run this benchmark in CI, since this was another trivial typo.

### CODE REVIEW

File: chia/cmds/keys_funcs.py
Comment (arvidn) [inline chia/cmds/keys_funcs.py:290]: the variables `public_key` and `signature` already have types, mypy doesn't allow shadowing the previous variables with new ones of different types

File: chia/util/cached_bls.py
Comment (arvidn) [inline chia/util/cached_bls.py:45]: `pair()` is commutative. this unifies the member function we use to be called on `G2Element`

File: chia/util/cached_bls.py
Comment (arvidn) [inline chia/util/cached_bls.py:50]: `pairings` is a list of *optional* `G2Element`s, but this function ensures they're all set, so we cast it to the correct return value

File: tests/blockchain/test_blockchain.py
Comment (arvidn) [inline tests/blockchain/test_blockchain.py:2077]: args is `List[bytes]`

File: tests/clvm/test_puzzles.py
Comment (arvidn) [inline tests/clvm/test_puzzles.py:37]: `key_gen` does not depend on the scheme. this is the only place in out code base where we refer to `BasicSchemeMPL`. Everywhere else we use `AugSchemeMPL`.

File: tests/pools/test_pool_rpc.py
Comment (arvidn) [inline tests/pools/test_pool_rpc.py:405]: `owner_sk` is a pair of (secret_key, index). This test was meant to compare the secret keys

File: tests/util/benchmark_cost.py
Comment (arvidn) [inline tests/util/benchmark_cost.py:116]: this seems like an old typo, it was introduced in the first commit of this file, Tue Mar 24 23:03:31 2020

File: tests/util/network_protocol_data.py
Comment (arvidn) [inline tests/util/network_protocol_data.py:63]: `blspy` does provide a constructor for `G1Element` and `G2Element` that takes bytes, I think it's a bit clearer to use the `from_bytes()` function

**Comment by arvidn** [comment]
```
tests/util/benchmark_cost.py (0.0%): Missing lines 116
```
This is a benchmark. We probably don't run it on CI, but even if we did, we don't instrument benchmarks for coverage

File: chia/cmds/keys_funcs.py
Comment (altendky) [reply chia/cmds/keys_funcs.py:290]: Do the `.from_bytes()` methods not hint the proper return type?  The `: G1Element` doesn't seem needed for clarity etc.

File: chia/util/cached_bls.py
Comment (altendky) [reply chia/util/cached_bls.py:50]: Maybe it would be better to have a separate list hinted properly such that mypy can help us enforce this?

File: tests/blockchain/test_blockchain.py
Comment (altendky) [reply tests/blockchain/test_blockchain.py:2077]: Did mypy catch this?  Or should be explicitly `: List[bytes]` to make sure it is enforced.

File: tests/pools/test_pool_rpc.py
Comment (altendky) [reply tests/pools/test_pool_rpc.py:405]: Wouldn't have caught the error here, but a not-tuple return would be nice.  Not for this PR though.

File: chia/cmds/keys_funcs.py
Comment (arvidn) [reply chia/cmds/keys_funcs.py:290]: good point. They are not necessary, I added them to explain to myself how these are new variables with different types

File: tests/pools/test_pool_rpc.py
Comment (arvidn) [reply tests/pools/test_pool_rpc.py:405]: yeah, I'm not entirely confident how this is related to `blspy` types, but it seems that untyped variables are "viral", so even comparing `Tuple[PrivateKey, int]` to `PrivateKey` doesn't trigger mypy, because `PrivateKey` is untyped

Comment (altendky) [reply tests/pools/test_pool_rpc.py:405]: We have strict equality enabled for the strict files, but not for the files in the mypy exceptions list.  But yes, if one thing is an `Any` then...  almost anything you do with it is `Any` as well and mypy won't complain much.

https://github.com/Chia-Network/chia-blockchain/blob/f0cb47baefbab6f0e94bb60dd8df9985bf4079dd/mypy.ini.template#L17

**Comment by arvidn** [comment]
@altendky when I run this locally, I get the following errors (and I don't understand why):
```
benchmarks/utils.py:76: error: Returning Any from function declared to return "G2Element"  [no-any-return]
chia/simulator/block_tools.py:545: error: Returning Any from function declared to return "G2Element | None"  [no-any-return]
Found 2 errors in 2 files (checked 811 source files)
```
but it seems to pass on CI

File: chia/rpc/wallet_rpc_api.py
Comment (arvidn) [inline chia/rpc/wallet_rpc_api.py:2257]: @matt-o-how This line is not exercised by any test, and has a (fatal) typo. I don't think it's reasonable to hold up this fix by adding a test. Do you want to vouch for this change?

File: chia/full_node/full_node.py
Comment (arvidn) [inline chia/full_node/full_node.py:1297]: apparently there's no test covering this line. I think it's trivially correct and warrants an exception. In the long run this function should probably be pulled out as a free function and unit tested

File: chia/rpc/wallet_rpc_api.py
Comment (arvidn) [reply chia/rpc/wallet_rpc_api.py:2257]: matt approves

File: install.sh
Comment (altendky) [inline install.sh:341]: If this is truly temporary then I guess we can let it stay separate, but the fewer pip calls the better it can handle itself.  I know we already have two and that's not good.

File: chia/util/cached_bls.py
Comment (altendky) [inline chia/util/cached_bls.py:50]: Maybe it would be better to have a separate list hinted properly such that mypy can help us enforce this?

File: chia/util/cached_bls.py
Comment (altendky) [reply chia/util/cached_bls.py]: reraised after modification

**Review by altendky — COMMENTED** [review]
I'm not excited about adding type stubs here for other packages, but I'm also not driving the future of these libs.  The asserts in the runtime code aren't great, but are also in keeping with our practices.

Other than above and my couple inline comments I'm ok with this.  Someone familiar with the actual code changes should review them specifically.

File: install.sh
Comment (arvidn) [reply install.sh:341]: I have a PR in the works that replaces blspy, but it's not ready yet.

File: chia/util/cached_bls.py
Comment (arvidn) [reply chia/util/cached_bls.py:50]: I went for `cast` to avoid making changes to the original intention. But I'll add a separate list to make the types more explicit

**Review by rostislav — APPROVED** [review]
Looks good

**Review by paninaro — APPROVED** [review]
lgtm

**Pull request merged**
